Refactor Observers

danthorpe opened the pull request:
This is to address the points raised in #152 by @jshier. Plus a few things I've noticed subsequently while fixing bugs for v2.6.

- [x] Supports Will/Did Finish
        The reasoning behind this, is that typically, observers want to know when the operation did *actually* finish. This means that `XCTAssertTrue(operation.finished)` should pass. However... `GroupOperation` and its like, will often want to use this observer (or a similar event) to add new operations to the queue. Hence, it used to be posted before the state change. Now, we just have two observers, one before and one after.

- [x] Renames methods to a better Swift style
        For example, instead of `operationDidStart` it is `didStartOperation`. Changed consistently throughout including in `GroupOperation`, which now provides `willFinishOperation` instead of `operationDidFinish` - note that this takes the change above into account too.       

- [x] Adds and calls a method for did attach to operation.
        The base protocol defines `didAttachToOperation` and in an extension, we have a default implementation which currently just logs a `.Verbose` message. In a custom observer, this can be over-ridden as you like.

- [x] Make thread safe & remove asserts from `Operation` when adding an observer.

danthorpe commented:
@jshier - I've got some tests to write, but I think the changes here should make the use of Operations a lot more flexible. I'll probably merge this over the weekend, but gonna be largely directed by your feedback on this - so if you could let me know what you think, and if there is anything more to do, that would be great!

danthorpe commented:
Okay, I'm going to merge this now, as it has quite a few breaking changes over 2.6, and so is going to be hard to stop getting out of sync if I leave it.

jshier commented:
I'll give this a try right now.

jshier commented:
Looks great so far. I'll take a closer look at the changes but my initial implementation of an `UIActivityIndicatorView` extension to observe an `Operation` works well. I'll work up to some more advanced implementations perhaps tomorrow. 

A few small notes: 
1. I was disappointed that the default `Block` observers didn't offer the ability to use `didAttach`, but I can see why keeping them simple is a priority. I was able to create my own block-based observers using it just find.
2. The `BlockType`s used by the `BlockObserver`s could use parameter labels, so when I autocomplete them I don't have to fill them in myself. Also, you might be able to define some of them in the protocols (I'm not sure), so other implementers have access to the definitions and labels.

I'll let you know if I have more feedback later.

danthorpe commented:
Hi @jshier - thanks for trying this out. Did you have to re-write much code? I added `@available(*, unavailable, renamed="etc etc")` annotations where possible - so hopefully Xcode offered to fix things for you.

Perhaps rather than an extension on `UIActivityIndicatorView` a bespoke observer which receives one would be useful to add to the project?

Regarding your notes:

1. Yeah, I can fix this, although I'm thinking that instead of accepting more blocks in an initializer, we can have mutable properties `var didAttachToOperation: (Operation -> Void)? = .None` *and* `var willDetachFromOperation: (Operation -> Void)? = .None` where the latter property would be called after the observer's last event is performed.

    For example - lets say you attach an `OperationDidStartObserver` - after the operation starts, if this observer does not conform to any of the other observation protocols, then it can be deallocated - but we can inform it before that happens. This observer could then be used to potentially attach different observers to the operation.

2. Yeah, this is no problem to fix either.

I will do another ticket later today.

danthorpe commented:
See #195 for pt. 2
